Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
-transitiveClosure core edges = runST $ do
-    list <- mapM (\node -> newSTRef Nothing >>= \ref -> return (node, ref)) core
-    let visited = Map.fromList list
-    mapM_ (collect visited) core
-    list' <- mapM (\(node, ref) -> readSTRef ref >>= \val -> return (node, val)) list
-    return [ (node, nodes) | (node, Just nodes) <- list' ]
-  where
-    collect :: Map a (STRef s (Maybe (Set a))) -> a -> ST s (Set a)
-    collect links node = do
-        case Map.lookup node links of
-            -- attempting to visit a node which wasn't given to us
-            Nothing  -> return Set.empty
-            Just ref -> readSTRef ref >>= \t -> case t of
-                -- the node has already been visited
-                Just calc -> return calc
-                Nothing   -> do
-                    -- Mark the node as visited with results pending. If a cycle
-                    -- brings us back to collect the same node, it will yield
-                    -- an empty list. This breaks the cycle for whatever node
-                    -- was visited first.
-                    --
-                    -- There are smarter algorithms that can
-                    -- get transitive closures of cyclic graphs, but cycles are
-                    -- highly pathological for the types of graphs we'll be
-                    -- traversing, so no need to worry.
-                    writeSTRef ref $ Just Set.empty
-                    let outEdges = edges node
-                    collected <- fmap Set.unions $ mapM (collect links) outEdges
-                    let connected = Set.union (Set.fromList outEdges) collected
-                    writeSTRef ref $ Just connected
-                    return connected
 
 ------------------------------------ Utility
--- For cases when, if a Map or Set is empty, it's as good as nothing at all.
-keepMap :: Ord k => Map k a -> Maybe (Map k a)
-keepMap con = if Map.null con then Nothing else Just con
 
-keepMaps :: (Ord k, Ord k') => (Map k a, Map k' b) -> Maybe (Map k a, Map k' b)
-keepMaps con@(c, c') = if Map.null c && Map.null c' then Nothing else Just con
+----------------------------Graph Utility----------
+suc :: RevDeps -> Vertex -> [Vertex]
+suc g v = g Arr.! v
 
-keepSet :: Ord a => Set a -> Maybe (Set a)
-keepSet con = if Set.null con then Nothing else Just con
+outdeg :: RevDeps -> Vertex -> Int
+outdeg r = length . suc r
 
---------------------------------------------------------------------------------
--- State
---
--- Last but agnostic of other ranking schemes,
--- methods for manipulating the global state.
+insEdges :: [(NodeId, [NodeId])] -> RevDeps -> RevDeps
+insEdges edges revdeps =
+    Arr.accum union revdeps edges
+
+delEdges :: [(NodeId, [NodeId])] -> RevDeps -> RevDeps
+delEdges edges revdeps =
+    Arr.accum (\\) revdeps edges
+
+revSize :: RevDeps -> Int
+revSize rev = snd $ Arr.bounds rev
+
+arrDouble :: RevDeps -> RevDeps
+arrDouble rev =
+    Gr.buildG (0, 2*revSize rev) (Gr.edges rev)
```

Comment:
This is not how immutable graphs work. Doubling the size helps nothing, since updating the graph has to copy anyway. That is, insEdges is already O(n), and if needs to add more nodes then that can be done in one pass at the same time, using `accumArray` with the new graph bounds.